Simplified/AxisService/NYPLAxisService.swift

```diff
@@ -73,18 +81,211 @@ class NYPLAxisService: NSObject {
+    group.wait()
+    guard let packageEndpoint = packageEndpoint else {
+      NYPLErrorLogger.logError(
+        nil,
+        summary: "Failed to generate container endpoint")
+      delegate?.failDownloadWithAlert(forBook: self.book)
+      return
+    }
+    
+    let packageURL = self.dedicatedWriteURL.appendingPathComponent(packageEndpoint)
+    guard
+      let data = try? Data(contentsOf: packageURL),
+      let xml = NYPLXML(data: data)
+      else {
+        NYPLErrorLogger.logError(
+          nil,
+          summary: "Failed to generate NYPLXML")
+        delegate?.failDownloadWithAlert(forBook: self.book)
+        return
+    }
     
+    let axisXML = NYPLAxisXML(xml: xml)
+    let hrefs = axisXML.findRecursivelyInAttributes(AxisHelper.hrefKey)
+    
+    for href in hrefs {
+      let endpath: String
+      if let pathPrefix = packagePathPrefix {
+        endpath = "\(pathPrefix)\(href)"
+      } else {
+        endpath = href
+      }
+      
+      let linkURL = baseURL
+        .appendingPathComponent(endpath)
+      
+      let writeURL = self.dedicatedWriteURL
+        .appendingPathComponent(endpath)
+      
+      group.enter()
+      downloadItem(from: linkURL, at: writeURL, group: group)
+    }
   }
   
+  // MARK: - Helper Methods
+  private func downloadItem(from url: URL,
+                            at writeURL: URL,
+                            group: DispatchGroup) {
+    
+    let downloader = NYPLContentDownloader(
+      networkExecutor: NYPLNetworkExecutor.shared)
+    
+    downloader.downloadContent(from: url) { (result) in
+      switch result {
+      case .success(let data):
+        do {
+          try NYPLAssetWriter.writeAsset(data, atURL: writeURL)
+          group.leave()
+        } catch {
+          NYPLErrorLogger.logError(
+            error,
+            summary: "Failed writing item to \(writeURL)")
+          
+          group.leave()
+          self.delegate?.failDownloadWithAlert(forBook: self.book)
+        }
+      case .failure(let error):
+        NYPLErrorLogger.logError(
+          error,
+          summary: "Failed downloading item from \(url)")
+        
+        group.leave()
+        self.delegate?.failDownloadWithAlert(forBook: self.book)
+      }
+    }
+  }
+  
+  /// We're going with lazy initialization instead of a computed property here because this variable is not
+  /// needed until the `container.xml` file has been downloaded. And since  it's going to produce the
+  /// same output every time, making it a computed property would be an overkill.
+  lazy private var packageEndpoint: String? = {
+    let containerURL = dedicatedWriteURL
+      .appendingPathComponent(AxisHelper.containerFileName)
+    
+    guard
+      let data = try? Data(contentsOf: containerURL),
+      let xml = NYPLXML(data: data)
+      else {
+        return nil
+    }
+    
+    return NYPLAxisXML(xml: xml)
+      .findRecursivelyInAttributes(AxisHelper.fullPathKey)
+      .first
+  }()
```

Comment:
Also, this way we ensure that NYPLAxisService does not violate Single Responsibility Principal

Reply:
Yeah, I think moving this and the one below to a separate class is a good idea.

Reply:
Done. I'm waiting for the entire feature to be finished before writing tests because chances are we might need to make more modifications in order to read a book. Once that's taken care of, we can write tests.